src/basilisp/core.lpy

```diff
@@ -6017,12 +6019,23 @@
                   (apply list (map fn-arity-with-destructuring body))
 
                   :else
-                  body)]
-    (as-> arities $
-      (cond->> $ name (cons name))
-      (cons 'fn* $)
-      (cond-> $
-        (meta &form) (with-meta (meta &form))))))
+                  body)
+        fn-decl (as-> arities $
+                  (cond->> $ name (cons name))
+                  (cons 'fn* $)
+                  (cond-> $
+                    (meta &form) (with-meta (meta &form))))
+
+        fmeta (merge (meta fn-decl) (meta name))
```

Comment:
Is this the correct order for how it is done in Clojure? I am indifferent towards the order, but I'd rather not have another ticket later just because it's wrong.

Reply:
To clarify, I believe you are asking about the merging order: should the fn `name` metadata override the `form` (`fn-decl`) metadata? Specifically, should the fn `name` metadata key `:decorators` take precedence over the `:decorators` key in the `form` metadata?

From a quick survey:
1. `defn` behavior: Both Clojure and Basilisp pass metadata to the `fn` function `name`.
2. Clojure  `fn` behavior: The [`fn` definition](https://github.com/clojure/clojure/blob/fb22fd778a272b034684a4ee94509552b46ee8a9/src/clj/clojure/core.clj#L4560) uses metadata keys from the `params` vector only for  `:pre` and `:post` conditions.

Thus based on the above, there’s no defined precedence in Clojure between fn `form` and `name` metadata. This leaves the choice up to us.

For the `fn` form, there are three potential places for the `:decorators` metadata key:
1. The fn `form` itself, i.e. `^{:decorators [...]} (fn ...)`).
2. The  fn `name` (if provided), i.e. `(fn ^{:decorators [...]} name ...)`.
3. The fn `params`, `(fn ^{:decorators [...]} [param1 ...] ...)

Since `defn` passes its metadata to the fn `name` (option 2), this seems like a strong candidate for support. Conceptually, `:decorators` are tied to the function, so supporting the `form` metadata key (option 1) also makes sense.

Conceptually to me, the `:decorators` is linked to the function, so I think support for the key in the `form` metadata should also be supported.

If both are supported, my view is:

- Specific overrides general: The function `name` metadata key should take precedence over the `form` metadata key.
- This aligns with the idea of specificity: metadata inside the `fn` form is more targeted than metadata outside it (but I think you can also equally argue in the opposite direction 😅)

Supporting `params` metadata (option 3) would also be possible, but it complicates precedence rules unnecessarily.

I prefer supporting both `form` and `name` metadata keys, with precedence given to the fn `name` key. Alternatively, for simplicity, we could support only the `form` metadata key.

What is your view?

Thanks

Reply:
> I prefer supporting both `form` and `name` metadata keys, with precedence given to the fn `name` key. Alternatively, for simplicity, we could support only the form metadata key.

I agree that we should support `form` and `name` with preference for `name` metadata.